src/Analysis/Ast/Impl/Analyzer/PythonAnalyzer.cs

```diff
@@ -281,19 +283,89 @@ private void StartAnalysis(IDependencyChainNode<PythonAnalyzerEntry> node, int v
                 var analysis = new DocumentAnalysis((IDocument)module, version, walker.GlobalScope, walker.Eval);
 
                 (module as IAnalyzable)?.NotifyAnalysisComplete(analysis);
-                node.Value.TrySetAnalysis(analysis, version, _syncObj);
+                entry.TrySetAnalysis(analysis, version);
+                node.Commit();
 
-                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) complete in {(stopWatch.Elapsed - startTime).TotalMilliseconds} ms.");
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) is completed in {(stopWatch.Elapsed - startTime).TotalMilliseconds} ms.");
             } catch (OperationCanceledException oce) {
-                node.Value.TryCancel(oce, version, _syncObj);
+                node.Value.TryCancel(oce, version);
+                node.Skip();
+                var module = node.Value.Module;
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) is canceled.");
             } catch (Exception exception) {
-                node.Value.TrySetException(exception, version, _syncObj);
+                var module = node.Value.Module;
+                node.Value.TrySetException(exception, version);
+                node.Commit();
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) has failed.");
```

Comment:
Forgive another nitpick, but given the other two, "analysis of _ failed" is more consistent, this one is even more minor though.

Should this print elapsed time as well?

Reply:
Elapsed time will be inaccurate at that point, cause building/catching exception is a long operation by itself.

Reply:
Fair enough.